### PR TITLE
[topgen] Support for external alert routing

### PR DESCRIPTION
**Pull request by Razer6**
This PR implements the external alert routing according to the [RFC](https://docs.google.com/document/d/1LGQjasiB6WJ1dL42hTKbzScmEmRQDXurdG6n8ORtJQk/edit?tab=t.0#heading=h.ly4mwumf0y6l).

### CODE REVIEW

File: hw/top_earlgrey/templates/toplevel.sv.tpl
Comment (rswarbrick) [inline hw/top_earlgrey/templates/toplevel.sv.tpl:51]: Why does this import need moving? I think you're using the name for e.g. the `outgoing_alert_*` signals with a complete path?

File: hw/top_earlgrey/templates/toplevel.sv.tpl
Comment (rswarbrick) [inline hw/top_earlgrey/templates/toplevel.sv.tpl:494]: Untested, but maybe this would repeat less?
```python
w = len(block.alerts)
if 'outgoing_alert' in m:
  outgoing_alert = m['outgoing_alert']
  lo = outgoing_alert_idx[outgoing_alert]
else:
  lo = alert_idx
slice = f"{lo+w-1}:{lo}"
```

File: util/topgen/merge.py
Comment (rswarbrick) [inline util/topgen/merge.py:903]: Probably cleaner:
```python
        alert_group = module.get('outgoing_alert')
        if alert_group is not None:
```

File: util/topgen/merge.py
Comment (rswarbrick) [inline util/topgen/merge.py:905]: I think this can just be:
```python
                outgoing_lpg_dict[lpg_name] = outgoing_num_lpg[alert_group]
```
?

Comment (rswarbrick) [inline util/topgen/merge.py:905]: I think I'd probably avoid using `defaultdict`. If this is the only place we might be incrementing for the first time, maybe it makes sense to do something like this here?
```python
                outgoing_num_lpg.setdefault(alert_group, 0)
```
Then `outgoing_num_lpg` can just be initialised with `{}`.

File: util/topgen/merge.py
Comment (rswarbrick) [inline util/topgen/merge.py:915]: Again, I think you can just do:
```python
                lpg_dict[lpg_name] = num_lpg
```

File: util/topgen/merge.py
Comment (rswarbrick) [inline util/topgen/merge.py:921]: Can't we get rid of `num_lpg` and just use `len(lpg_dict)`?

File: util/topgen/merge.py
Comment (rswarbrick) [inline util/topgen/merge.py:910]: The flow of information through the `clock` variable is pretty confusing. Can we just add it as an argument to `append_to_lpg_dict`?

File: util/topgen/merge.py
Comment (rswarbrick) [inline util/topgen/merge.py:936]: I realise that this is copied from the lines above, but can't we just do this?
```python
        for alert_group, alerts in top['outgoing_alert'].items():
            for alert in alerts:
                if alert['module_name'] == module['name']:
                    alert['lpg_name'] = lpg_name
                    alert['lpg_idx'] = outgoing_lpg_dict[module['outgoing_alert']][lpg_name]
```

**Review by andreaskurth — COMMENTED** [review]
Thanks for this work @Razer6. The underlying RFC is currently being reviewed, so I won't approve this PR yet. The changes in the PR seem to be aligned with the RFC, and in that sense they look good to me.

Please also add the RFC doc as Markdown to this PR. (You may want to wait with that until it's fully approved, though.)

File: hw/top_darjeeling/templates/toplevel.sv.tpl
Comment (Razer6) [reply hw/top_darjeeling/templates/toplevel.sv.tpl:413]: That one is defined in [here](https://github.com/lowRISC/opentitan/pull/25179/files#diff-e4870f59f3d7b3080fabf9d5216c5a87104b49a1b98b1b9bb55b3157d1d4e612L679)

File: hw/top_earlgrey/templates/toplevel.sv.tpl
Comment (Razer6) [reply hw/top_earlgrey/templates/toplevel.sv.tpl:51]: I think that one can be removed. I though originally it was needed for the signal definitions.

File: util/topgen/merge.py
Comment (Razer6) [reply util/topgen/merge.py:905]: Is there any reason not to use it?

File: util/topgen/merge.py
Comment (Razer6) [reply util/topgen/merge.py:921]: No I don't think that would work. We need a mapping of LPG name to the index in the vector to be able to later associate the alerts to that.

File: hw/top_darjeeling/templates/toplevel.sv.tpl
Comment (rswarbrick) [reply hw/top_darjeeling/templates/toplevel.sv.tpl:413]: Oh! I see: the definitions are things like `clkmgr_aon_clocks.clk_io_div4_peri` and this wants to get `io_div4_peri` here. (A singleton but I hadn't twigged we weren't splitting on spaces)

If I've got it right, this is essentially the `clk` signal in `merge.py`, line 900? Is there any way that the generated json file can preserve more of the structure? If not, maybe it makes sense to be more explicit about how we're splitting things here? Since this split happens several times in the code, maybe it needs to be in a helper function.

File: util/topgen/merge.py
Comment (rswarbrick) [reply util/topgen/merge.py:915]: I think I agree that we should keep the dictionary, but I'm reasonably sure that `num_lpg` is always equal to `len(lpg_dict)`, so I'm proposing that we get rid of the `num_lpg` variable.

File: util/topgen/merge.py
Comment (Razer6) [reply util/topgen/merge.py:893]: That is refactored to use `clk`

File: util/topgen/merge.py
Comment (Razer6) [reply util/topgen/merge.py:915]: Ahhh, sorry. I misread that. Got it now. Yeah, as we construct the dict, we can use its length to get the mapping index.

File: util/topgen/merge.py
Comment (Razer6) [reply util/topgen/merge.py]: That one is now gone.

**Comment by Razer6** [comment]
@rswarbrick I think I addressed all your comments. Could you take another look?

**Comment by Razer6** [comment]
> Please also add the RFC doc as Markdown to this PR. (You may want to wait with that until it's fully approved, though.)

@andreaskurth I added an issue to track the documentation of that.

File: hw/top_darjeeling/templates/toplevel.sv.tpl
Comment (Razer6) [reply hw/top_darjeeling/templates/toplevel.sv.tpl]: In the last commit, I simplified that a lot I think.

**Comment by andreaskurth** [comment]
> @andreaskurth I added an issue to track the documentation of that.

For reference, that would be https://github.com/lowRISC/opentitan/issues/25449

**Comment by andreaskurth** [comment]
AUTHORIZE CHANGES: hw/top_earlgrey/rtl/autogen/top_earlgrey.sv

Just whitespaces; no functional changes.

**Comment by vogelpi** [comment]
CHANGE AUTHORIZED: hw/top_earlgrey/rtl/autogen/top_earlgrey.sv

Just whitespaces; no functional changes.

File: hw/top_darjeeling/templates/toplevel.sv.tpl
Comment (rswarbrick) [reply hw/top_darjeeling/templates/toplevel.sv.tpl]: I really like it. Thanks!

File: util/topgen/merge.py
Comment (rswarbrick) [inline util/topgen/merge.py:913]: Sorry to be annoying, but I'd probably suggest checking with `is not None` to avoid silly things happening if we accidentally have an alert group with an empty name.

File: util/topgen/merge.py
Comment (rswarbrick) [reply util/topgen/merge.py]: Looks good, thanks!

**Review by rswarbrick — APPROVED** [review]
This looks good to me, thanks. One nitty comment about empty strings but it needn't really gate this going in.

**Comment by rswarbrick** [comment]
CHANGE AUTHORIZED: hw/top_earlgrey/rtl/autogen/top_earlgrey.sv

As Andreas said, this is just a couple of blank lines: no functional change.

**Pull request merged**
